presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduSplitManager.java

```diff
@@ -59,4 +81,58 @@ public ConnectorSplitSource getSplits(
                 throw new IllegalArgumentException("Unknown splitSchedulingStrategy: " + splitSchedulingStrategy);
         }
     }
+
+    private static CompletableFuture<?> whenCompleted(DynamicFilter dynamicFilter)
+    {
+        if (dynamicFilter.isAwaitable()) {
+            return dynamicFilter.isBlocked().thenCompose(ignored -> whenCompleted(dynamicFilter));
+        }
+        return NOT_BLOCKED;
+    }
+
+    private static class KuduDynamicFilteringSplitSource
+            implements ConnectorSplitSource
+    {
+        private final CompletableFuture<?> dynamicFilterFuture;
+        private final CompletableFuture<ConnectorSplitSource> splitSourceFuture;
+
+        private KuduDynamicFilteringSplitSource(
+                CompletableFuture<?> dynamicFilterFuture,
+                CompletableFuture<ConnectorSplitSource> splitSourceFuture)
+        {
+            this.dynamicFilterFuture = requireNonNull(dynamicFilterFuture, "dynamicFilterFuture is null");
+            this.splitSourceFuture = requireNonNull(splitSourceFuture, "splitSourceFuture is null");
+        }
+
+        @Override
+        public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHandle partitionHandle, int maxSize)
+        {
+            return splitSourceFuture.thenCompose(splitSource -> splitSource.getNextBatch(partitionHandle, maxSize));
+        }
+
+        @Override
+        public void close()
+        {
+            if (!dynamicFilterFuture.cancel(true)) {
+                splitSourceFuture.thenAccept(ConnectorSplitSource::close);
+            }
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            if (splitSourceFuture.isDone()) {
+                if (splitSourceFuture.isCompletedExceptionally()) {
+                    return false;
```

Comment:
I don't think it can complete exceptionally

Reply:
If isFinished() is called after close(), and dynamicFilterFuture cancellation in close() succeeded, splitSourceFuture is completed exceptionally.

Reply:
It should return `true` then here

Reply:
In this case, splitSource is not created since getSplitSource() is not triggered. I think return false is more reasonable. 
 isFinished() return false after close() should not be a problem. In FixedSplitSource,  isFinished() aslo can return false after close().